sdk/src/OpenTelemetry/Processor/Batch.hs

```diff
@@ -198,35 +196,70 @@ buildExport m =
   )
 
 
--- | Exitable forever loop
-loop :: Monad m => ExceptT e m a -> m e
-loop = liftM (either id id) . runExceptT . forever
-
-
 data ProcessorMessage = Flush | Shutdown
 
 
+-- note: [Unmasking Asyncs]
+--
+-- It is possible to create unkillable asyncs. Behold:
+--
+-- ```
+-- a <- uninterruptibleMask_ $ do
+--     async $ do
+--         forever $ do
+--             threadDelay 10_000
+--             putStrLn "still alive"
+-- cancel a
+-- ```
+--
+-- The prior code block will never successfully cancel `a` and will be
+-- blocked forever. The reason is that `cancel` sends an async exception to
+-- the thread performing the action, but the `uninterruptibleMask` state is
+-- inherited by the forked thread. This means that *no async exceptions*
+-- can reach it, and `cancel` will therefore run forever.
+--
+-- This also affects `timeout`, which uses an async exception to kill the
+-- running job. If the action is done in an uninterruptible masked state,
+-- then the timeout will not successfully kill the running action.
+
 {- |
  The batch processor accepts spans and places them into batches. Batching helps better compress the data and reduce the number of outgoing connections
  required to transmit the data. This processor supports both size and time based batching.
 -}
 batchProcessor :: MonadIO m => BatchTimeoutConfig -> Exporter ImmutableSpan -> m Processor
 batchProcessor BatchTimeoutConfig {..} exporter = liftIO $ do
   batch <- newIORef $ boundedMap maxQueueSize
-  workSignal <- newEmptyMVar
-  worker <- async $ loop $ do
-    req <-
-      liftIO $
-        timeout (millisToMicros scheduledDelayMillis) $
-          takeMVar workSignal
-    batchToProcess <- liftIO $ atomicModifyIORef' batch buildExport
-    res <- liftIO $ Exporter.exporterExport exporter batchToProcess
-
-    -- if we were asked to shutdown, quit cleanly after this batch
-    -- FIXME: this could lose batches if there's more than one in queue?
-    case req of
-      Just Shutdown -> throwE res
-      _ -> pure ()
+  workSignal <- newEmptyTMVarIO
+  shutdownSignal <- newEmptyTMVarIO
+  let workerAction = do
+        delay <- registerDelay (millisToMicros scheduledDelayMillis)
+        req <- atomically $ do
+          msum
+            [ const Flush <$> do
+                continue <- readTVar delay
+                check continue
```

Comment:
`STM`'s `<|>` is defined as [`orElse`](https://hackage.haskell.org/package/base-4.17.0.0/docs/GHC-Conc-Sync.html#v:orElse), and `msum` is `asum` is `foldr (<|>) empty`. 

`orElse` haddock:

> Compose two alternative STM actions (GHC only).
>
> If the first action completes without retrying then it forms the result of the [orElse](https://hackage.haskell.org/package/base-4.17.0.0/docs/GHC-Conc-Sync.html#v:orElse). Otherwise, if the first action retries, then the second action is tried in its place. If both actions retry then the [orElse](https://hackage.haskell.org/package/base-4.17.0.0/docs/GHC-Conc-Sync.html#v:orElse) as a whole retries.